iwyu.cc

```diff
@@ -2718,7 +2652,83 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
   // void fn(const S<Class>& s) { // Forward declarations are sufficient here.
   //   (void)s.t; // Full 'Class' type is needed due to template instantiation.
   // }
-  void ReportTplSpecComponentTypes(const TemplateSpecializationType*) = delete;
+  void ReportTplSpecComponentTypes(const TemplateSpecializationType*,
+                                   const set<const Type*>& blocked_types) =
+      delete;
+
+  const set<const Type*>& GetBlockedTypes() const = delete;
+
+  void ReportTypeUseInternal(SourceLocation used_loc, const Type* type,
+                             const char* comment,
+                             const set<const Type*>& provided_types) {
```

Comment:
I had decided to divide terminology into "provided" and "blocked" because types may be "blocked" not only because they are "provided by author". E.g., I intended to use `blocked_types_` inside `InstantiatedTemplateVisitor` also for blocking types on the instantiation side which are used on the declaration side, when (and if) implementing improved dependent type handling in the future. On the other hand, such types can also be considered as "provided by template declaration"... Not sure.

I like "author types" and "ignored types" terminology.

Reply:
I see. I was expecting the division to be the inverse, i.e. `ReportTypeUse` would use "blocked" and callers would use "provided", as there are more reasons to block than types being provided by a template defn. But yeah, this is all very fuzzy. :)

Reply:
Reasonable, thanks. `ReportTypeUseInternal` parameter should better be called `blocked_types`.
> being provided by a **template** defn

Just in case, for clarity: the current PR treats only types in _typedef and type alias_ definitions.

"Definition-side types" term is also good, my only concern is that it looks too long.

Reply:
Agree that it's too long, I just wanted to throw it out there in case it would inspire something better yet.